File: LeetCode/916/Solution.cpp
class Solution {
public:
    bool compa(string a, string b) {
        int idx1 = 0, idx2 = 0;
        while(a.size() > idx1 && b.size() > idx2) {
            if(a[idx1] == b[idx2]) {
                idx1++;
                idx2++;
            } else {
                idx1++;
            }
        }
        return idx2 == b.size();
    }
    vector<string> wordSubsets(vector<string>& words1, vector<string>& words2) {
        vector<string> ret;
        vector<int> count(26, 0);
        
        for(string word : words2) {
            vector<int> temp(26, 0);
            for(char c : word) {
                temp[c-'a']++;
            }
            for(int i=0;i<26;++i) {
                count[i] = max(count[i], temp[i]);
            }
        }
        
        string comp = "";
        for(int i=0;i<26;++i) {
            for(int j=0;j<count[i];++j) {
                comp += 'a'+i;
            }
        }
        
        vector<string> temp;
        for(string& word : words1) {
            string t = word;
            sort(t.begin(), t.end());
            temp.push_back(t);
        }

        for(int i=0;i<words1.size();++i) {
            if(compa(temp[i], comp)) {
                ret.push_back(words1[i]);
            }
        }
        
        return ret;
    }
};

File: LeetCode/916/Solution.java
class Solution {
    public List<String> wordSubsets(String[] words1, String[] words2) {
        int[] diff = new int[26];
        for(String word : words2) {
            int[] count = new int[26];
            for(int i=0;i<word.length();++i) {
                count[word.charAt(i)-'a']++;
            }
            for(int i=0;i<26;++i) {
                diff[i] = Math.max(diff[i], count[i]);
            }
        }

        List<String> result = new ArrayList<>(); 
        for(String word : words1) {
            int[] count = new int[26];
            for(int i=0;i<word.length();++i) {
                count[word.charAt(i)-'a']++;
            }
            boolean flag = true;
            for(int i=0;i<26;++i) {
                if(diff[i] > count[i]) {
                    flag = false;
                    break;
                }
            }
            if(flag) {
                result.add(word);
            }
        }
        return result;
    }
}
